src/python/pants/reporting/async_workunit_handler.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import threading
+from contextlib import contextmanager
+from typing import Any, Callable, Iterator, Optional
+
+
+DEFAULT_REPORT_INTERVAL_SECONDS = 10
+
+
+class AsyncWorkunitHandler:
+  def __init__(self, scheduler: Any, callback: Optional[Callable], report_interval_seconds: float = DEFAULT_REPORT_INTERVAL_SECONDS):
+    self.scheduler = scheduler
+    self.report_interval = report_interval_seconds
+    self.callback = callback
+    self._thread_runner = None
+
+  def start(self):
+    self._thread_runner = _InnerHandler(self.scheduler, self.callback, self.report_interval)
```

Comment:
I think I'll change this so that `_InnerHandler` always assumes a callback exists, and `AsyncWorkunitHandler` is in charge of ensuring that the inner class doesn't get created if the callback isn't set.